### PR TITLE
LUCENE-10266 Move nearest-neighbor search on points to core

### PR DESCRIPTION
**Pull request by shahrs87**
# Description

Please provide a short description of the changes you're making with this pull request.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/lucene/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [ ] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.

### CODE REVIEW

**Comment by shahrs87** [comment]
@jpountz I have created this PR as per your suggestion in LUCENE-10266 jira. I have made the following assumptions. Please correct me if needed.
1. I have deleted LatLonPointPrototypeQueries class since there is no other sandobx query in that class. Should I keep the empty class ?
2. I see we have FloatPointNearestNeighbor implementation in sandbox which is similar to NearestNeighbor. Do I need to remove FloatPointNearestNeighbor from sandbox and move it to lucene/core ?
3. I have added this change to API Changes section in CHANGES.txt Please correct if it belongs somewhere else.

Thank you.

**Comment by shahrs87** [comment]
@jpountz  @iverase @mikemccand  Can you guys please help me review this patch ? Tagging Ignacio and Michael since you showed interest in the corresponding jira. Thank you.

File: lucene/core/src/java/org/apache/lucene/document/LatLonPoint.java [hunk 1]
```diff
@@ -362,4 +377,72 @@ public static Query newDistanceFeatureQuery(
...
     return query;
   }
+
+  /**
+   * Finds the {@code n} nearest indexed points to the provided point, according to Haversine
+   * distance.
+   *
+   * <p>This is functionally equivalent to running {@link MatchAllDocsQuery} with a {@link
+   * LatLonDocValuesField#newDistanceSort}, but is far more efficient since it takes advantage of
+   * properties the indexed BKD tree. Currently this only works with {@link Lucene90PointsFormat}
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/document/LatLonPoint.java:387]: this point that it only works with Lucene90PointsFormat is no longer true, can you remove it?

File: lucene/core/src/java/org/apache/lucene/search/NearestNeighbor.java [hunk 2]
```diff
@@ -31,12 +31,8 @@
...
 import org.apache.lucene.util.SloppyMath;
 
-/**
- * KNN search on top of 2D lat/lon indexed points.
- *
- * @lucene.experimental
- */
-class NearestNeighbor {
+/** KNN search on top of 2D lat/lon indexed points. */
+public class NearestNeighbor {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/NearestNeighbor.java:35]: Can we keep it pkg-private if we have it in oal.document? This is especially important as we're promoting this class to core.

File: lucene/core/src/java/org/apache/lucene/document/LatLonPoint.java (hunk 1 above)
Comment (shahrs87) [reply lucene/core/src/java/org/apache/lucene/document/LatLonPoint.java]: Removed in the latest revision.

File: lucene/core/src/java/org/apache/lucene/search/NearestNeighbor.java (hunk 2 above)
Comment (shahrs87) [reply lucene/core/src/java/org/apache/lucene/search/NearestNeighbor.java]: Yes. Changed in the latest revision.

**Comment by shahrs87** [comment]
@jpountz  Thank you for the feedback. I have addressed your comments in the latest revision. I have one question. To make changes in the CHANGES.txt file, will this change go in `API changes` section or `Other`. Please advise.

File: lucene/core/src/java/org/apache/lucene/document/LatLonPoint.java [hunk 3]
```diff
@@ -362,4 +375,71 @@ public static Query newDistanceFeatureQuery(
...
+   *
+   * @param searcher IndexSearcher to find nearest points from.
+   * @param field field name. must not be null.
+   * @param latitude latitude at the center: must be within standard +/-90 coordinate bounds.
+   * @param longitude longitude at the center: must be within standard +/-180 coordinate bounds.
+   * @param n the number of nearest neighbors to retrieve.
+   * @return TopFieldDocs containing documents ordered by distance, where the field value for each
+   *     {@link FieldDoc} is the distance in meters
+   * @throws IllegalArgumentException if the underlying PointValues is not a {@code
+   *     Lucene60PointsReader} (this is a current limitation), or if {@code field} or {@code
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/document/LatLonPoint.java:400]: we removed this limitation that `Lucene60PointsReader` is required

File: lucene/core/src/java/org/apache/lucene/document/NearestNeighbor.java [hunk 4]
```diff
@@ -220,7 +216,7 @@ public Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
   }
 
   /** Holds one hit from {@link NearestNeighbor#nearest} */
-  static class NearestHit {
+  public static class NearestHit {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/document/NearestNeighbor.java:219]: can it remain pkg-private too?

**Review by jpountz — COMMENTED** [review]
Thanks, it looks almost ready. Maybe as a next step we could try to make this functionality correct on multi-valued fields.

**Comment by shahrs87** [comment]
@jpountz Addressed all of your previous comments. Please review again. Thank you.

**Comment by jpountz** [comment]
Thanks @shahrs87 !

**Pull request merged**